### PR TITLE
Go modules for apis/cni and apis/cpi

### PR DESCRIPTION
**Pull request by akutz**
### What this PR does / why we need it
This patch adds support for treating the API packages for CNI and CPI as distinct Go modules so remote repositories can import these modules without inheriting the entire tanzu-framework dependency graph.

### Which issue(s) this PR fixes

Fixes `NA`

### Describe testing done for PR
I built all of the project's Go modules:

```bash
$ find . -name go.mod -print0 | \
  xargs -0 dirname | \
  xargs -I% -n1 bash -c 'cd %; echo; echo "building $(pwd)"; time go build ./...; cd - 2>&1 >/dev/null'

building /Users/akutz/Projects/tanzu-framework

real	0m55.398s
user	8m0.626s
sys	1m41.372s

building /Users/akutz/Projects/tanzu-framework/apis/cpi

real	0m1.892s
user	0m1.939s
sys	0m3.881s

building /Users/akutz/Projects/tanzu-framework/apis/cni

real	0m4.307s
user	0m4.804s
sys	0m8.457s

building /Users/akutz/Projects/tanzu-framework/addons

real	0m12.477s
user	0m14.798s
sys	0m13.605s

building /Users/akutz/Projects/tanzu-framework/addons/pinniped/post-deploy

real	0m10.863s
user	0m10.646s
sys	0m9.936s

building /Users/akutz/Projects/tanzu-framework/addons/pinniped/config-controller

real	0m8.444s
user	0m9.629s
sys	0m9.122s

building /Users/akutz/Projects/tanzu-framework/hack/tools
go: warning: "./..." matched no packages

real	0m0.033s
user	0m0.012s
sys	0m0.015s

building /Users/akutz/Projects/tanzu-framework/hack/packages/kbld-image-replace

real	0m2.972s
user	0m3.043s
sys	0m3.320s

building /Users/akutz/Projects/tanzu-framework/hack/packages/package-tools

real	0m1.644s
user	0m1.476s
sys	0m1.893s

building /Users/akutz/Projects/tanzu-framework/pkg/v1/providers/tests

real	0m1.231s
user	0m1.235s
sys	0m2.322s
```

### Release note

```release-note
package-based-lcm: Changed the apis/cni and apis/cpi packages to be their own Go modules so other projects can import those APIs without pulling in this project's entire dependency graph.
```

### PR Checklist

- [x] Squash the commits into one or a small number of logical commits
      
- [x] Use good commit [messages](https://github.com/vmware-tanzu/tanzu-framework/blob/main/CONTRIBUTING.md)
- [x] Ensure PR contains terms all contributors can understand and links all contributors can access

### Additional information
cc @vijaykatam @sidharthsurana @jmoroski

### CODE REVIEW

**Comment by akutz** [comment]
Hi @jmoroski / @vijaykatam ,

One odd thing I noticed is the Pinniped config-controller does not reference the parent module by relative path but by a pinned version. Is this on purpose? Ex. https://github.com/vmware-tanzu/tanzu-framework/blob/ba2b0e1f0679bf7ad2593b62e39496fb4e1f3cbb/addons/pinniped/config-controller/go.mod#L12

This is anachronistic to how other modules in this repository reference the parent module, ex. https://github.com/vmware-tanzu/tanzu-framework/blob/ba2b0e1f0679bf7ad2593b62e39496fb4e1f3cbb/addons/go.mod#L6

cc @sidharthsurana @mayankbh

File: go.mod [hunk 1]
```diff
@@ -2,6 +2,10 @@ module github.com/vmware-tanzu/tanzu-framework
 
 go 1.17
 
+replace github.com/vmware-tanzu/tanzu-framework/apis/cni => ./apis/cni
```
Comment (vijaykatam) [inline go.mod:5]: nitpick: Could combine replace statements into one block

**Review by vijaykatam — COMMENTED** [review]
Thanks for making this change. I think pinniped could refer to relative path as well. Will ping @ankeesler

**Comment by akutz** [comment]
> Thanks for making this change. I think pinniped could refer to relative path as well. Will ping @ankeesler

Thanks. I already pinged him in the comment up above, but this way he'll definitely see it.

File: go.mod (hunk 1 above)
Comment (akutz) [reply go.mod]: Fixed.

**Review by vijaykatam — APPROVED** [review]
LGTM.

**Comment by akutz** [comment]
> > Thanks for making this change. I think pinniped could refer to relative path as well. Will ping @ankeesler
> 
> Thanks. I already pinged him in the comment up above, but this way he'll definitely see it.

FWIW, we resolved offline to separate the changes.

**Comment by sidharthsurana** [comment]
LGTM

**Review by lubronzhan — APPROVED** [review]

**Pull request merged**